### ISSUE TITLE
Is 4store still active?

### ISSUE THREAD

**Issue by er5678ihgfrt5y6ftgtyr** [4store/4store#143]
Is 4store still actively developed and maintained?

**Comment by berezovskyi** [comment]
Ha, what is so secret in this question you had to create a throwaway account?

I have kept an eye on the project for some time and its not active for sure. But the original devloper, @swh, is very responsive when issues are filed and especially when PRs are submitted. I have been putting away the mini-project to get familiar with the 4store code base myself.

I am actually quite interested to know what are the main problems of the 4store at the moment, @swh?

And also, would it be possible for you, @swh, to mark some of the open issues you think are relatively easy for the newcomers as `help-needed` or `low-hanging-fruit` etc.?

**Comment by er5678ihgfrt5y6ftgtyr** [comment]
> Ha, what is so secret in this question you had to create a throwaway account?

ahah not a secret, it's just that I prefer to support free software and therefore don't have a GitHub account. The 4store website doesn't exist, and I couldn't find any mailing list, so I just created this random account.

Anyway... I'm looking for a free/libre RDF triplestore and there seems to be only a bunch of them. I'm not interested in supporting any of those stores that come with shady oss/proprietary dual licenses, *and* I'd also like to dodge Java if I can. So, ultimately what I ended up with is

- [RedStore](https://www.aelius.com/njh/redstore/): seems dead
- 4Store: seems dead

Is there any actively maintained, libre, possibly non-java, and possibly scalable RDF store?

**Comment by swh** [comment]
Honestly, it's not actively maintained - there are a few entities using it in production, but in well-established projects, I would guess.

I think it lags behind the SPARQL standard a bit, as Garlik created another proprietary SPARQL store after 4store, that never made it into the free software world.

We lost the website because the company that bough Garlik acquired the domain, and turned off the server that was running 4store.org - someone archived the content though.

@berezovskyi good idea about tagging issues, I'll make a note to do that. The codebase is quite complex though, there's probably only 4 or 5 people in the world that really understand it.

**Comment by er5678ihgfrt5y6ftgtyr** [comment]
Is there any free/libre (no proprietary dual license) that you would recommend?

**Comment by swh** [comment]
I don't know to be honest. We use 4store for processing RDF data we receive in my current company, but it's not in the production critical path, so I didn't spend much time looking for alternatives.

**Comment by TravisWhitaker** [comment]
Although 4store development has ground to a halt, it's nonetheless quite stable. At my previous job at a large pharmaceutical company we used a small 4store cluster for data integration; to the best of my knowledge that team is still using 4store today. @er5678ihgfrt5y6ftgtyr if you don't need the extra functionality stipulated by SPARQL 1.1, I wouldn't hesitate to give 4store a shot, depending on the application you have in mind.

**Comment by er5678ihgfrt5y6ftgtyr** [comment]
@swh 

- can 4store do inferencing?
- is there any chance that development will start again if people use it?

**Comment by er5678ihgfrt5y6ftgtyr** [comment]
can 4store do rdfs inferencing?

**Comment by er5678ihgfrt5y6ftgtyr** [comment]
is there any chance that development will start again if people use 4store? I'm asking because *all* rdf stores suck, and those who suck less are released with shitty oss/proprietary dual licenses (of course the oss license is treated as an "evaluation" license, so I don't care about that software).
4store doesn't seem to suck, plus it's written in C rather than Java and it's also a native store (= doesn't use 3rd party rdbms/nosql stores, unless I'm missing something). But.... it's no longer maintainer, and I'm not sure I want to use a software in the long-term that will never receive any updates or bug fixes...

**Comment by swh** [comment]
4store does not to RDFS reasoning, but 4sr does https://github.com/msalvadores/4sr/wiki

I'm sure development would start again if there was a need, but there are some significant gaps between SPARQL 1.1 and 4store's implementation, which won't be filled easily. I suspect most people would rather use a Java-based engine.

**Comment by er5678ihgfrt5y6ftgtyr** [comment]
> I suspect most people would rather use a Java-based engine.

well... Java is often bloated, and requires to install a huge dependency like the JDK (even ignoring the fact that Oracle JDK is **not** free/libre). I've tried Titan for example (well, it's a graph store not really an rdf store), and it downloads hundreds if not thousands of dependencies, hundreds of megabytes, and I'm not even sure if it's downloading any blobs (since it downloads from Oracle too). Furthermore, most of them seem to use some non-native back-end, and the benchmarks that I can find online seem to point that those Java stores don't really scale that well...

**Comment by berezovskyi** [comment]
@er5678ihgfrt5y6ftgtyr, I think you are not being constructive here. Apache/EPL software is not free enough, Java requires a JDK (OpenJDK works great for us in many projects), GPL/commercial license is not libre enough (BTW, if you want libre-libre triplestore, forking the GPL'ed project is the best you can do).

I am developer of one linked-data related Java project licensed under EPL. I can't understand your reasoning. Java code we write is not super awesome libre stuff. It's just free (from patents as well) code that GETS. THINGS. DONE. And writing good code takes time and effort. If you'd come to our project asking about features, I'd tell you that I would be very happy to do the code review and merge your changes.

Also, from my experience in that project, I learned that getting the release process going is the key. I think that we can look into #117 and maybe get @jonassmedegaard's and @swh guidance.

**Comment by kjetilk** [comment]
I would love to see 4store restarted. I think it is very well done and certainly, since there are many bridges to other programming languages from C, it could play an important part in a larger ecosystem. 

But then, the problem I have, and many have, is time to contribute. I have none... :-) Would you be able to commit time to releasing it @swh? Are there other options that could be worthwhile trying to revive it?

**Comment by swh** [comment]
Realistically I can only spare a few hours a week, and more importantly I've not written more than a few lines of C for 5+ years :-/

**Comment by er5678ihgfrt5y6ftgtyr** [comment]
> Apache/EPL software is not free enough

wait... I don't have a problem with the Apache license. Where did I say that? I only have a problem with Oracle and their licensing.

> Java requires a JDK (OpenJDK works great for us in many projects)

I don't hate Java a priori; if there are only databases written in Java, I'll use them. But can we agree that if a much less bloated solution exist, I'd rather choose less bloat than more bloat?

> GPL/commercial license is not libre enough

the problem here are companies, not "not libre enough". I don't have any issue with copyleft or permissive licenses. The problem is the business model of these companies that release their "free" or "community" version only as "evaluation" (bait) copies, and then work against you to lure you into their "commercial", non-free version and try to make you dependent on other non-free components. Their interest is to make their free version deficient of important feature on purpose, hoping you'll jump in their proprietary-bandwagon. Plus, I don't want to contribute or support a piece of software that's going to be resold as proprietary alongside other blobware. However, if the **exact** copy is dual licensed oss/commercial I'll probably consider using it.

**Comment by berezovskyi** [comment]
@swh few hours is a lot of time for reviewing pull requests, thank you!

@er5678ihgfrt5y6ftgtyr sorry, I might have misunderstood some of your statements. I do not fully agree on the bloat, it's a rather vague statement. True, there is memory management overhead etc. but for the DB server that is supposed to use GBs of RAM and TBs of storage, I don't see 200MB RAM overhead for the JVM and 100MB deployment WAR file as much of a bloat because it's relatively insignificant. There are other problems in the JVM-based solutions, but that's something else than "bloat". In re: GPL, you can fork the base store and improve it with the rest of the features (and once you fork it, you no longer have to sign the CLA for commercial redistribution). But I mainly use Fuseki and can't comment on more advanced stores like Blazegraph etc. (seems like Blazegraph is GPL2 OSS but the CUDA extension is proprietary and I don't see it as anything terribly bad that prohibits you from evaluating it).

tl;dr:

* the project is active as long as we want to put some energy into it
* there are some correctness bugs (I will label those once I get the repo access) and SPARQL 1.1 support is far from complete
* @swh is ready to spend a few hours a week on it, which can mean ~5 PR reviews or so

Shall we close it and move the tech discussion in the rest of the bug tracker (starting with #117)?

**Comment by berezovskyi** [comment]
Also, I think there is some newsletter that we can use for the general discussion? Is https://groups.google.com/forum/#!forum/4store-support controlled by @swh?

**Comment by swh** [comment]
Yes, I control it.

**Comment by er5678ihgfrt5y6ftgtyr** [comment]
Does it work without a Google account?

**Comment by swh** [comment]
Yes

**Comment by berezovskyi** [comment]
continued in #154
